### DISCUSSION TITLE
toggle between banks with cable like ManyAddresses would

### DISCUSSION THREAD

**Discussion by gringow831** [tttapa/Control-Surface#699] [General]
Hi Pieter, 
Thought of creating a new discussion specific to this.
I am trying to create a multibank option that will have about 8 combinations. For now, I wrote 3 and the variables but can't make it work and don't understand why.
I failed when trying the ManyAddresses with CABLE so instead as each part would have a combination of cables I thought of creating every bank manually and a variable to change between them. So far it has 3 "banks" with 3 faders to try making it work although it will have 8 MCU PBPotentiometers and Buttons combining different cables. In an ideal world it would work in the same way as IncrementDecrementSelector having the option to Wrap::Wrap or Wrap::Clamp.
I am using a Teensy 4.0.

It does seem like a very long way of writing the code with a variable for 8 banks but can't think of more ways of doing this as my coding knowledge is limited. 

```cpp
#include <Control_Surface.h>

USBMIDI_Interface midi;

Button bankUp = 3;
Button bankDown = 2;

/*
IncrementDecrementSelector<4> bankbuttons = {
  bank0, {bankUp, bankDown},  Wrap::Wrap,
};
*/

//channels 5 (CABLE_3), 6 (CABLE_3), 2 (CABLE_2)
PBPotentiometer fadersbank_1[] = { 
  {14, {MCU::VOLUME_5, CABLE_3}},
  {15, {MCU::VOLUME_6, CABLE_3}},
  {17, {MCU::VOLUME_2, CABLE_2}},
};

//channels 7 (CABLE_2), 8 (CABLE_2), 2 (CABLE_1)
PBPotentiometer fadersbank_2[] = {
  {14, {MCU::VOLUME_7, CABLE_2}},
  {15, {MCU::VOLUME_8, CABLE_2}},
  {16, {MCU::VOLUME_1, CABLE_1}},
};

//channels 1 (CABLE_4), 2 (CABLE_4), 3 (CABLE_2)
PBPotentiometer fadersbank_3[] = { 
  {14, {MCU::VOLUME_1, CABLE_4}},
  {15, {MCU::VOLUME_2, CABLE_4}},
  {17, {MCU::VOLUME_3, CABLE_2}},
};

//PBPotentiometer fadersbank_4
//PBPotentiometer fadersbank_5
//PBPotentiometer fadersbank_6
//PBPotentiometer fadersbank_7
//PBPotentiometer fadersbank_8

NoteButton noteButton[] = {
  { 5, MCU::BANK_LEFT,},
  { 7, MCU::BANK_RIGHT,},
};
  
void setup() {
  RelativeCCSender::setMode(MACKIE_CONTROL_RELATIVE);

  Control_Surface.begin();
  //delay(20000);
  Control_Surface.sendNoteOn(MCU::BANK_RIGHT, 127);
  Selector<2 >::disable(fadersbank_2);
 Selector<2 >::disable(fadersbank_3);
 }

void loop() { 
   Control_Surface.loop();
  //button to go up the banks with specific cables - fadersbank 1 to 2
    if (bankUp.update() == Button::Falling) {
   PBPotentiometer::disable(fadersbank_1);
   PBPotentiometer::disable(fadersbank_3;
   PBPotentiometer::enable(fadersbank_2);
  }  
//button to go up the banks with specific cables - fadersbank 2 to 3 
if (bankUp.getState() == Button::Falling) {
   PBPotentiometer::disable(fadersbank_2);
   PBPotentiometer::disable(fadersbank_1);
   PBPotentiometer::enable(fadersbank_3);
  } 
/* wrap the fadersbanks
   else if (bankUp.update() == Button::Falling) {
   PBPotentiometer::disable(fadersbank_3);
   PBPotentiometer::disable(fadersbank_2);
   PBPotentiometer::enable(fadersbank_1);
  } 
*/

  //button to go down the banks with specific cables
    if (bankDown.update() == Button::Falling) {
   PBPotentiometer::disable(fadersbank_1);
   PBPotentiometer::disable(fadersbank_3);
   PBPotentiometer::enable(fadersbank_2);
  }  
//button to go up the banks with specific cables - fadersbank 2 to 3 
if (bankDown.getState() == Button::Falling) {
   PBPotentiometer::disable(fadersbank_2);
   PBPotentiometer::disable(fadersbank_3);
   PBPotentiometer::enable(fadersbank_1);
  } 

  }
   
}
```

**Comment by gringow831** [comment]
This is something I tried too to see if I could make it work. 

```cpp
Bankable::ManyAddresses::PBPotentiometer<3> faderbanks[] = {
  {bank0, 14, {{MCU::VOLUME_5, CABLE_3}, {MCU::VOLUME_7, CABLE_2}, {MCU::VOLUME_1, CABLE_4}}},
};
```

**Reply by tttapa** [reply]
> This is something I tried too to see if I could make it work.
> 
> ```c++
> Bankable::ManyAddresses::PBPotentiometer<3> faderbanks[] = {
>   {bank0, 14, {{MCU::VOLUME_5, CABLE_3}, {MCU::VOLUME_7, CABLE_2}, {MCU::VOLUME_1, CABLE_4}}},
> };
> ```

What is it that doesn't work exactly? Do you get a compile time error, runtime error, does it send on the wrong channel or cable, does it not send anything at all?

**Reply by gringow831** [reply]
error: could not convert 

Arduino: 1.8.19 (Mac OS X), TD: 1.56, Board: "Teensy 4.0, MIDIx4, 600 MHz, Faster, US English"

 warning: 'CS::MCU::VU' has a field 'CS::MCU::VU::decayTimer' whose type uses the anonymous namespace
 class VU : public MatchingMIDIInputElement<MIDIMessageType::CHANNEL_PRESSURE,
       ^
many_addresses_cable:10: error: could not convert '{bank, 14, {{CS::MCU::VOLUME_5, CS::CABLE_3}, {CS::MCU::VOLUME_7, CS::CABLE_2}, {CS::MCU::VOLUME_1, CS::CABLE_4}}}' from '<brace-enclosed initializer list>' to 'CS::Bankable::ManyAddresses::PBPotentiometer<3u>'
 };
 ^
Using library Control-Surface-master at version 2.0.0 in folder: /Users/gringow/Documents/Arduino/libraries/Control-Surface-master 
Using library SPI at version 1.0 in folder: /Applications/Teensyduino.app/Contents/Java/hardware/teensy/avr/libraries/SPI 
could not convert '{bank, 14, {{CS::MCU::VOLUME_5, CS::CABLE_3}, {CS::MCU::VOLUME_7, CS::CABLE_2}, {CS::MCU::VOLUME_1, CS::CABLE_4}}}' from '<brace-enclosed initializer list>' to 'CS::Bankable::ManyAddresses::PBPotentiometer<3u>'


Can't compile the program when using cables with manyaddresses

**Reply by tttapa** [reply]
You need another pair of braces:
```cpp
Bankable::ManyAddresses::PBPotentiometer<3> faderbanks[] {
  {
    bank0,
    14, 
    {{
      {MCU::VOLUME_5, CABLE_3},
      {MCU::VOLUME_7, CABLE_2},
      {MCU::VOLUME_1, CABLE_4},
    }}
  },
};
```

**Reply by gringow831** [reply]
I have been counting braces for days and couldn't find the error! Thank you so much for the quick response! It does work now and even better don't need the whole variable now

**Reply by tttapa** [reply]
It's a bit counterintuitive, but this is the explanation:
```cpp
Bankable::ManyAddresses::PBPotentiometer<3> faderbanks[] { // start of list
  { // PBPotentiometer constructor
    bank0,
    14, 
    { // Array constructor
      { // start of list
        { // MIDIChannelCN constructor
          MCU::VOLUME_5, CABLE_3,
        },
        { // MIDIChannelCN constructor
          MCU::VOLUME_7, CABLE_2,
        },
        { // MIDIChannelCN constructor
          MCU::VOLUME_1, CABLE_4,
        },
      } // end of list
    } // end of Array constructor
  } // end of PBPotentiometer constructor
}; // end of list
```
If each element of the inner list has only a single argument, you don't need braces around them, and you can omit the braces for the Array constructor as well. If the elements are more complicated and require braces, the placing of the omitted braces becomes ambiguous and the compiler complains.

**Reply by gringow831** [reply]
That makes sense. Can't believe the answer was so simple
